docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,844 @@
+type ControlPlaneProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   ControlPlaneProviderSpec   `json:"spec,omitempty"`
+   Status ControlPlaneProviderStatus `json:"status,omitempty"`
+}
+
+// ControlPlaneProviderSpec defines the desired state of ControlPlaneProvider
+type ControlPlaneProviderSpec struct {
+   ProviderSpec `json:",inline"`
+}
+
+// ControlPlaneProviderStatus defines the observed state of ControlPlaneProvider
+type ControlPlaneProviderStatus struct {
+   ProviderStatus `json:",inline"`
+}
+
+// InfrastructureProvider is the Schema for the InfrastructureProviders API
+type InfrastructureProvider struct {
+   metav1.TypeMeta   `json:",inline"`
+   metav1.ObjectMeta `json:"metadata,omitempty"`
+
+   Spec   InfrastructureProviderSpec   `json:"spec,omitempty"`
+   Status InfrastructureProviderStatus `json:"status,omitempty"`
+}
+
+// InfrastructureProviderSpec defines the desired state of InfrastructureProvider
+type InfrastructureProviderSpec struct {
+   ProviderSpec `json:",inline"`
+}
+
+// InfrastructureProviderStatus defines the observed state of InfrastructureProvider
+type InfrastructureProviderStatus struct {
+   ProviderStatus `json:",inline"`
+}
+```
+
+Below you can find details about `ProviderSpec`,  `ProviderStatus`, which is
+shared among all the provider types - Core, Bootstrap, ControlPlane, and
+Infrastructure.
+
+```golang
+// ProviderSpec defines the desired state of the Provider.
+type ProviderSpec struct {
+   // Version indicates the provider version.
+   // +optional
+   Version string `json:"version,omitempty"`
+
+   // Manager defines the properties that can be enabled on the controller manager for the provider.
+   // +optional
+   Manager ManagerSpec `json:"manager,omitempty"`
+
+   // Deployment defines the properties that can be enabled on the deployment for the provider.
+   // +optional
+   Deployment DeploymentSpec `json:"deployment,omitempty"`
+
+   // SecretName is the name of the Secret providing the configuration
+   // variables for the current provider instance, like e.g. credentials.
+   // Such configurations will be used when creating or upgrading provider components.
+   // +optional
+   SecretName *string
+
+   // FetchConfig determines how the operator will fetch the components and metadata for the provider.
+   // If nil, the operator will try to fetch components according to default
+   // settings embedded in the operator and in clusterctl.
+   // +optional
+   FetchConfig *FetchConfiguration `json:"fetchConfig,omitempty"`
+
+   // Paused prevents the operator from reconciling the provider. This can be
+   // used when doing an upgrade or move action manually.
+   // +optional
+   Paused bool `json:"paused,omitempty"`
+}
+
+// ManagerSpec defines the properties that can be enabled on the controller manager for the provider.
+type ManagerSpec struct {
```

Comment:
I think it is worth to put some focus on the UX around feature flags because they are currently managed as a sort of exception given that currently, they are the only element of a manager spec that can be influenced by the users during init (using variables).
Also, in the current clusterctl, the same feature flag variable applies to many providers

Reply:
I included a property for `FeatureGates` in the `ManagerSpec` and clarified its behavior in the validation rules for `ContainerSpec.Args`.